frame/stake-tracker/migrations/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+								*current = current.saturating_add(stake);
+							}
+
+							iterations = iterations.saturating_add(Weight::from_ref_time(1));
+							nominator_state.last_key = next_key;
+							leftover_weight =
+								max_iterations.saturating_sub(iterations).saturating_sub(
+									Weight::from_ref_time(approval_stakes.len() as u64),
+								);
+
+							if leftover_weight.all_lte(Weight::default()) {
+								// We ran out of weight, also taking into account writing
+								// approval_stakes here. Save the state and bail.
+								Self::set_nominator_state(nominator_state.clone());
+
+								return (approval_stakes, leftover_weight)
+							}
+						},
+						None => {
+							log!(warn, "an un-decodable nomination detected.");
+							break
+						},
+					};
+				}
+			}
+
+			let mut validator_state = Self::validator_state();
+
+			if !validator_state.done {
+				// each iteration = 1 read + 2 reads = 3 reads
+				while let Some(next_key) =
+					sp_io::storage::next_key(validator_state.last_key.as_ref())
+				{
+					if !next_key.starts_with(&validator_state.prefix) {
+						// Nothing else to iterate. Save the state and bail.
+						validator_state.done = true;
+						Self::set_validator_state(validator_state.clone());
+						break
+					}
+
+					// Should be safe due to the check above.
+					let mut account_raw =
+						next_key.strip_prefix(validator_state.prefix.as_slice()).unwrap();
+
+					// TODO: not sure if this works, gotta see. Should be safe as all the keys in
+					// Validators map are AccountId. If not - let it fail.
+					let who = T::AccountId::decode(&mut account_raw).unwrap();
+					let stake = Pallet::<T>::slashable_balance_of(&who);
+					let current = approval_stakes.entry(who).or_default();
+					*current = current.saturating_add(stake);
+					iterations = iterations.saturating_add(Weight::from_ref_time(1));
+					validator_state.last_key = next_key;
+
+					leftover_weight = max_iterations
+						.saturating_sub(iterations)
+						.saturating_sub(Weight::from_ref_time(approval_stakes.len() as u64));
+
+					if leftover_weight.all_lte(Weight::default()) {
+						// We ran out of weight, also taking into account writing
+						// approval_stakes here. Save the state and bail.
+						Self::set_validator_state(validator_state.clone());
+
+						return (approval_stakes, leftover_weight)
+					}
+				}
+			}
+
+			(approval_stakes, leftover_weight)
+		}
+	}
+
+	impl<T: pallet_staking::Config + pallet_stake_tracker::Config> OnRuntimeUpgrade
+		for InjectValidatorsApprovalStakeIntoTargetList<T>
+	{
+		fn on_runtime_upgrade() -> Weight {
+			// We have to set this manually, because we need this migration to happen in order for
+			// the pallet to get all the data from staking-pallet.
+			let current = StorageVersion::new(1);
+			let onchain = Pallet::<T>::on_chain_storage_version();
+
+			if current == 1 && onchain == 0 {
+				let max_weight = T::BlockWeights::get().max_block;
```

Comment:
This function is only called once, I think what you actually want to do is to perform the migration `on_idle`. Since the state of this pallet is never read, this should be fine.

Reply:
We can't perform this on-idle, we need to make sure that the nominators/validators maps don't change, so we need to take all the weight to prevent this, unless I'm missing something